### PR TITLE
AEA-1177 Env vars support for agent configuration

### LINKED ISSUE #2049: Env file vars overwrite yaml config
**Describe the bug**
Currently, when env vars are specified in the yaml config then they are substituted on first load with their values.

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

Env vars support for agent configuration

## Fixes

#2049

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -344,51 +347,6 @@ def _create_or_update_from_json(
         """Create new config object or updates existing one from json data."""
         raise NotImplementedError  # pragma: nocover
 
-    @staticmethod
-    def _compare_data_to_pattern(
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:348]: moved where?

File: tests/test_configurations/test_manager.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+from unittest.mock import patch
+
+import pytest
+import yaml
+
+from aea.configurations.manager import AgentConfigManager
+
+
+agent_config_data = yaml.safe_load(
+    """
```
Comment (DavidMinarsch) [inline tests/test_configurations/test_manager.py:32]: move into data as a file?

File: tests/test_aea_builder.py [hunk 3]
```diff
@@ -466,20 +466,6 @@ def test_component_add_bad_dep():
         builder.add_component_instance(a_protocol)
 
 
-def test_find_component_failed():
```
[the remark this answers is not in the document]
Comment (solarw) [reply tests/test_aea_builder.py:469]: cause code moved to another module

File: aea/configurations/validation.py [hunk 4]
```diff
@@ -216,3 +248,55 @@ def required_fields(self) -> List[str]:
...
         return self._schema["required"]
+
+
+def validate_data_with_pattern(
+    data: dict,
+    pattern: dict,
+    excludes: Optional[List[Tuple[str]]] = None,
+    skip_env_vars: bool = False,
+) -> List[str]:
+    """Validate data dict with pattern dict for attributes present and type match."""
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/validation.py:259]: done

File: aea/configurations/validation.py [hunk 5]
```diff
@@ -142,7 +171,10 @@ def split_component_id_and_config(
 
     @classmethod
     def validate_component_configuration(
-        cls, component_id: ComponentId, configuration: Dict
+        cls,
+        component_id: ComponentId,
+        configuration: Dict,
+        env_vars_friendly: bool = False,
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/validation.py:177]: done

File: aea/configurations/manager.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    AgentConfig,
+    ComponentConfiguration,
+    ComponentId,
+    DEFAULT_AEA_CONFIG_FILE,
+    PackageType,
+)
+from aea.configurations.constants import PRIVATE_KEY_PATH_SCHEMA, VENDOR
+from aea.configurations.loader import ConfigLoader, load_component_configuration
+from aea.crypto.helpers import create_private_key, try_validate_private_key_path
+from aea.crypto.registries import crypto_registry
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/manager.py:41]: done

File: aea/configurations/base.py [hunk 7]
```diff
@@ -732,19 +699,6 @@ def _create_or_update_from_json(
 
         return instance
 
-    def update(self, data: Dict) -> None:
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/base.py:735]: it was an outdated code. actually, parent class update is do all we need.

File: aea/configurations/base.py (hunk 1 above)
Comment (solarw) [reply aea/configurations/base.py:348]: moved to validtion

File: tests/test_configurations/test_manager.py (hunk 2 above)
Comment (solarw) [reply tests/test_configurations/test_manager.py:32]: it used in one place only, it's easier to see it in file

File: aea/configurations/manager.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+                        identifier, sorted(crypto_registry.supported_ids)
+                    )
+                )
+
+        for identifier in crypto_registry.supported_ids:
+            config_private_key_path = aea_conf.private_key_paths.read(identifier)
+
+            if is_env_variable(config_private_key_path):
+                # skip env var definition
+                continue
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/manager.py:360]: better comment provided

**Pull request merged**